### PR TITLE
Refactored TableGen backends

### PR DESCRIPTION
**Pull request by Rot127**
Adds the refactored TableGen backends used by `auto-sync` (https://github.com/capstone-engine/capstone/pull/1949).

The refactor basically adds two printer classes. One for LLVM and one for Capstone. Emitter backends then call methods of one of those to emit the code.

For some emitter backends the `PrinterCapstone` class will generate additional mapping tables.

This design is not really sophisticated but works for our use case and is relatively easy to fix in case of conflicts during rebase. But it has some inherent problems when it comes to emitting anything else then the code generated by the emitter backends.

Unfortunately this is a problem from LLVM. See the discussion here (starting at comment from `nhaehnle`): https://reviews.llvm.org/D138323

Refactoring the emitter backends needs a lot of work and involvement from the LLVM folks, which there seems to be no motivation for currently.

### CODE REVIEW

File: llvm/utils/TableGen/InstrInfoEmitter.cpp [hunk 1]
```diff
@@ -922,250 +575,133 @@ void InstrInfoEmitter::run(raw_ostream &OS) {
...
-  OS << "\nextern const MCInstrDesc " << TargetName << "Insts[] = {\n";
+  PI.instrInfoEmitMCInstrDescHdr(TargetName);
   ArrayRef<const CodeGenInstruction*> NumberedInstructions =
     Target.getInstructionsByEnumValue();
 
   SequenceToOffsetTable<std::string> InstrNames;
-  unsigned Num = NumberedInstructions.size();
-  for (const CodeGenInstruction *Inst : reverse(NumberedInstructions)) {
+  // CAPSTONE: Do not reverse NumberedInstructions
+  // Our we access it via opcode. It must be ascending for this.
```
Comment (XVilka) [inline llvm/utils/TableGen/InstrInfoEmitter.cpp]: A typo?

**Review by XVilka — APPROVED** [review]

**Comment by reitowo** [comment]
Hope all auto-sync related PRs can be merged soon~

**Comment by Rot127** [comment]
@cnSchwarzer Me as well :D

Please be aware though, even if the ARM related PRs are merged the auto-sync work is not finalized yet. If you are looking into it for development purposes, please base your work on the PPC branches (https://github.com/capstone-engine/capstone/pull/2013 and https://github.com/Rot127/llvm-capstone/tree/tblgen_capstone_backends_ppc).
`auto-sync` will not be stable until `v5.1`.

**Comment by reitowo** [comment]
Thanks for notice! It seems I just come into this timing that capstone is switching architecture :) Looking forward to this.

**Comment by Rot127** [comment]
> It seems I just come into this timing that capstone is switching architecture 

The API for `disasm()` etc. will stay the same. But each arch module will have (sometimes significant) changes because they will be closer to LLVM (see the `Possibly breaking changes` list in https://github.com/capstone-engine/capstone/pull/1949).

**Comment by kabeor** [comment]
@Rot127 Plz pr to `auto-sync` branch, `release/x.x` is just for archive llvm code.

**Comment by Rot127** [comment]
@kabeor Done

**Comment by kabeor** [comment]
Great! Merged.

**Pull request merged**